tests/NuGet.Services.AzureSearch.FunctionalTests/BasicTests/V3SearchProtocolTests.cs

```diff
@@ -0,0 +1,322 @@
+        {
+            // Act
+            var result = await V3SearchAsync(new V3SearchBuilder() { Query = Constants.TestPackageId });
+
+            // Assert
+            Assert.True(result.TotalHits.HasValue);
+            Assert.True(result.TotalHits.Value > 0, $"Could not find test package {Constants.TestPackageId}");
+            Assert.True(result.Data.Count > 0, $"Could not find test package {Constants.TestPackageId}");
+
+            Assert.NotNull(result.AtContext);
+            Assert.True(result.AtContext.AtVocab == "http://schema.nuget.org/schema#");
+            Assert.False(string.IsNullOrEmpty(result.AtContext.AtBase));
+
+            // Assert that the package result whose Id is "BaseTestPackage" with Version "1.0.0"
+            // matches exactly what is expected.
+            var package = result.Data
+                .Where(p => p.Id == Constants.TestPackageId)
+                .Where(p => p.Version == Constants.TestPackageVersion)
+                .FirstOrDefault();
+
+            Assert.NotNull(package);
+            Assert.False(string.IsNullOrEmpty(package.AtId));
+            Assert.True(package.AtType == "Package");
+            Assert.False(string.IsNullOrEmpty(package.Registration));
+            Assert.True(package.Description == Constants.TestPackageDescription);
+            Assert.True(package.Summary == Constants.TestPackageSummary);
+            Assert.True(package.Title == Constants.TestPackageTitle);
+            Assert.True(package.Tags.Count() == 2);
+            Assert.True(package.Tags[0] == "Tag1");
+            Assert.True(package.Tags[1] == "Tag2");
+            Assert.True(package.Authors.Count() == 1);
+            Assert.True(package.Authors[0] == Constants.TestPackageAuthor);
+            Assert.True(package.TotalDownloads != default(long));
+            Assert.True(package.Versions.Count() == 1);
+            Assert.True(package.Versions[0].Version == "1.0.0");
+            Assert.True(package.Versions[0].Downloads != default(long));
+            Assert.False(string.IsNullOrEmpty(package.Versions[0].AtId));
+        }
+
+        [Theory]
+        [MemberData(nameof(TakeResults))]
+        public async Task TakeReturnsExactResults(int take)
+        {
+            var results = await V3SearchAsync(new V3SearchBuilder { Query = "json", Take = take });
+
+            Assert.NotNull(results);
+            Assert.True(results.TotalHits > take);
+            Assert.True(results.Data.Count == take, $"The search result did not return the expected {take} results");
+        }
+
+        [Fact]
+        public async Task SkipDoesSkipThePackagesInResult()
+        {
+            var searchTerm = "json";
+            var skip = 5;
+            var resultsWithoutSkip = await V3SearchAsync(new V3SearchBuilder { Query = searchTerm, Take = 10 });
+            var resultsWithSkip = await V3SearchAsync(new V3SearchBuilder { Query = searchTerm, Skip = skip, Take = 10 - skip });
```

Comment:
There is a small race condition here, results can change over time but I think it's acceptable

Reply:
you mean if the search ranking changes between line 101 and 102? That is an extremely miniscule possibility :)

Reply:
talked offline. The possibility of mismatch would be in cases where the search score is tied for the results, in such cases its possible the results might get swapped at the threshold. This is very unlikely to happen with azure search.